File: app-v3/src/test/java/com/khmelenko/lab/varis/viewmodel/TestSearchResultsViewModel.java
package com.khmelenko.lab.varis.viewmodel;

import com.khmelenko.lab.varis.RxJavaRules;
import com.khmelenko.lab.varis.repositories.search.SearchResultsViewModel;

import org.junit.Rule;

import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

/**
 * Testing {@link SearchResultsViewModel}
 *
 * @author Dmytro Khmelenko (d.khmelenko@gmail.com)
 */
public class TestSearchResultsViewModel {

    @Rule
    public RxJavaRules mRxJavaRules = new RxJavaRules();

    // TODO
//    @Inject
//    TravisRestClient mTravisRestClient;
//
//    private SearchResultsViewModel mSearchResultsViewModel;
//
//    private SearchResultsView mSearchResultsView;
//
//    @Before
//    public void setup() {
//        TestComponent component = DaggerTestComponent.builder().build();
//        component.inject(this);
//
//        mSearchResultsViewModel = spy(new SearchResultsViewModel(mTravisRestClient));
//        mSearchResultsView = mock(SearchResultsView.class);
//        mSearchResultsViewModel.attach(mSearchResultsView);
//    }
//
//    @Test
//    public void testStartRepoSearch() {
//        final String searchQuery = "test";
//        final List<Repo> responseData = new ArrayList<>();
//        when(mTravisRestClient.getApiService().getRepos(searchQuery)).thenReturn(Single.just(responseData));
//
//        mSearchResultsViewModel.startRepoSearch(searchQuery);
//        verify(mSearchResultsView).hideProgress();
//        verify(mSearchResultsView).setSearchResults(eq(responseData));
//    }
//
//    @Test
//    public void testStartRepoSearchFailed() {
//        final String searchQuery = "test";
//
//        final String errorMsg = "error";
//        Exception exception = new Exception(errorMsg);
//        when(mTravisRestClient.getApiService().getRepos(searchQuery)).thenReturn(Single.error(exception));
//
//        mSearchResultsViewModel.startRepoSearch(searchQuery);
//        verify(mSearchResultsView).hideProgress();
//        verify(mSearchResultsView).showLoadingError(eq(exception.getMessage()));
//    }
}

File: app-v3/src/test/java/com/khmelenko/lab/varis/viewmodel/TestSearchResultsViewModel.kt
package com.khmelenko.lab.varis.viewmodel

import android.arch.core.executor.testing.InstantTaskExecutorRule
import android.arch.lifecycle.Observer
import com.khmelenko.lab.varis.RxJavaRules
import com.khmelenko.lab.varis.dagger.DaggerTestComponent
import com.khmelenko.lab.varis.network.response.Repo
import com.khmelenko.lab.varis.network.retrofit.travis.TravisRestClient
import com.khmelenko.lab.varis.repositories.search.SearchResultsViewModel
import com.khmelenko.lab.varis.repositories.search.SearchState
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Single
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import javax.inject.Inject

/**
 * Testing [SearchResultsViewModel]
 *
 * @author Dmytro Khmelenko (d.khmelenko@gmail.com)
 */
class TestSearchResultsViewModel {

    @get:Rule
    var rxJavaRules = RxJavaRules()

    @get:Rule
    var testRule: TestRule = InstantTaskExecutorRule()

    @Inject
    lateinit var travisRestClient: TravisRestClient

    lateinit var searchResultsViewModel: SearchResultsViewModel

    private val stateObserver = mock<Observer<SearchState>>()

    @Before
    fun setup() {
        val component = DaggerTestComponent.builder().build()
        component.inject(this)

        searchResultsViewModel = SearchResultsViewModel(travisRestClient)

        searchResultsViewModel.state().observeForever(stateObserver)
    }

    @Test
    fun testStartRepoSearch() {
        val searchQuery = "test"
        val responseData = ArrayList<Repo>()
        whenever(travisRestClient.apiService.getRepos(searchQuery)).thenReturn(Single.just(responseData))

        searchResultsViewModel.startRepoSearch(searchQuery)

        verify(travisRestClient.apiService).getRepos(searchQuery)

        verify(stateObserver).onChanged(SearchState.Loading)
        verify(stateObserver).onChanged(SearchState.ReposList(responseData))
        verifyNoMoreInteractions(stateObserver)
    }

    @Test
    fun testStartRepoSearchFailed() {
        val searchQuery = "test"

        val errorMsg = "error"
        val exception = Exception(errorMsg)
        whenever(travisRestClient.apiService.getRepos(searchQuery)).thenReturn(Single.error(exception))

        searchResultsViewModel.startRepoSearch(searchQuery)

        verify(travisRestClient.apiService).getRepos(searchQuery)

        verify(stateObserver).onChanged(SearchState.Loading)
        verify(stateObserver).onChanged(SearchState.Error(errorMsg))
        verifyNoMoreInteractions(stateObserver)
    }
}
